libbeat/autodiscover/README.md

```diff
@@ -0,0 +1,260 @@
+
+Metricbeat and Filebeat support autodiscover based on hints from the provider. The hints system looks for hints in Kubernetes Pod annotations which have the prefix co.elastic.metrics. As soon as the container starts, Metricbeat/Filebeat will check if it contains any hints and launch the proper config for it. Hints tell Metricbeat/Filebeat how to get metrics for the given container.
+
+Example metricbeat configuration:
+```
+metricbeat.autodiscover:
+      providers:
+        - type: kubernetes
+          node: ${NODE_NAME}
+          hints.enabled: true
+```
+
+Example redis pod with right annotations:
+```
+apiVersion: v1
+kind: Pod
+metadata:
+  name: redis
+  annotations:
+    co.elastic.metrics/module: redis
+    co.elastic.metrics/metricsets: info
+    co.elastic.metrics/hosts: '${data.host}:6379'
+spec:
+  containers:
+  - name: redis
+    image: redis:5.0.4
+    command:
+      - redis-server
+      - "/redis-master/redis.conf"
+    env:
+    - name: MASTER
+      value: "true"
+    ports:
+    - containerPort: 6379
+    resources:
+      limits:
+        cpu: "0.1"
+    volumeMounts:
+    - mountPath: /redis-master-data
+      name: data
+    - mountPath: /redis-master
+      name: config
+  volumes:
+    - name: data
+      emptyDir: {}
+    - name: config
+      configMap:
+        name: example-redis-config
+        items:
+        - key: redis-config
+          path: redis.conf
+```
+
+##### How hints work
+
+Everything works the same as Autodiscover without LeaderElection until step 8.
+
+8. If there is no conditions in the template set by the user, the configs will be generated from [hints](https://github.com/elastic/beats/blob/master/libbeat/autodiscover/providers/kubernetes/kubernetes.go#L186).
+9. Wether hints are enabled or not is part of the [Kubernetes Provider struct](https://github.com/elastic/beats/blob/master/libbeat/autodiscover/providers/kubernetes/kubernetes.go#L121) builders field.
+10. GenerateHints[https://github.com/elastic/beats/blob/eff92354db783001880f4bade9f59942fca747ba/libbeat/autodiscover/builder/helper.go#L213] function looks into the event's annotations. A [hints map](https://github.com/elastic/beats/blob/eff92354db783001880f4bade9f59942fca747ba/libbeat/autodiscover/builder/helper.go#L226) is created with all hints and returned.
+11. From those hints, configs are [created](https://github.com/elastic/beats/blob/master/libbeat/autodiscover/builder.go#L97) in the same form as in `Autodiscover without LeaderElection` step 8.
+    They contain the same information as if they where set explicitly in the metricbeat configureation but actually derive from the pod annotations.
+12. Those configs are then [added](https://github.com/elastic/beats/blob/master/libbeat/autodiscover/providers/kubernetes/kubernetes.go#L197) in the event and gets published.
+13. The process after that is same as in `Autodiscover without LeaderElection` step 9 and onward.
+
+
+
+##  Autodiscover provider in Elastic Agent
```

Comment:
I would suggest moving this part under `elastic-agent` codebase so as to be close to the provider's code. Then you can cross-link it from libbeat's docs.